File: Google Docs (Document Writer)/Bad Design/Python/DocumentWriter.py

class DocumentWriter:
    def __init__(self):
        self._elements = []

    # methods
    def addtext(self, text):
        self._elements.append(text)

    def addimage(self, image):
        self._elements.append("[Image: " + image + " ]")

    def changetexttobold(self, startindex, endindex):
        for i in range(startindex, min(len(self._elements), endindex + 1)):
            element = self._elements[i]
            element = "**" + element + "**"
            self._elements[i] = element
    
    def changetexttoitellic(self, startindex, endindex):
        for i in range(startindex, min(len(self._elements), endindex + 1)):
            element = self._elements[i]
            element = "*" + element + "*"
            self._elements[i] = element
    
    def renderer(self):
        content = ""
        for element in self._elements:
            content += str(element) + "\n"
        return content
    
    def save(self):
        content = self.renderer()
        with open("document.txt", "w") as file:
            file.write(content)
        print("Document saved as document.txt")


if __name__=="__main__":    
    print("This module is not intended to be run directly.")
    doc = DocumentWriter()
    text1 = "Hello, this is a sample document. "
    text2 = "This document contains text and images. "
    image1 = "image1.png"
    doc.addtext(text1)
    doc.addtext(text2)
    doc.addimage(image1)
    doc.changetexttobold(0,1)
    doc.save()

File: Google Docs (Document Writer)/Bad Design/Java/DocumentWriter.java

import java.util.List;
import java.io.FileWriter;
import java.util.ArrayList;

class Documents{
    List<StringBuilder> elements = new ArrayList<>(); // List to hold document // line by line
    // Methods to manipulate document
    void addtext(StringBuilder text){
        elements.add(text);
    }
    void addimage(StringBuilder image){
        StringBuilder imagetext = new StringBuilder("[Image: " + image + " ]");
        elements.add(imagetext);
    }
    private String renderer(){
        StringBuilder renderedContent = new StringBuilder();
        for(StringBuilder element : elements){
            renderedContent.append(element).append("\n");
        }
        return renderedContent.toString();
    }
    void save(){
        String content = renderer();
        System.out.println("Saving document with content: " + content);
        try{
            FileWriter writer = new FileWriter("document.txt");
            writer.write(content);
            writer.close();
        }
        catch(Exception e){
            System.out.println("An error occurred while saving the document.");
        }
    }
    void changetexttobold(int startindex, int endindex){
        for(int i = startindex; i < Math.min(elements.size(),endindex+1); i++){
            StringBuilder element = elements.get(i);
            element.append("**").append(element).append("**");
            elements.set(i, element);
        }
    }
    void changetexttoitellic(int startindex, int endindex){
        for(int i = startindex; i < Math.min(elements.size(),endindex+1); i++){
            StringBuilder element = elements.get(i);
            element.append("*").append(element).append("*");
            elements.set(i, element);
        }
    }
}


public class DocumentWriter{
    public static void main(String args[]){
        System.out.println("This is a bad design example.");
        Documents doc = new Documents();
        StringBuilder text1 = new StringBuilder("Hello, this is a sample document. ");
        StringBuilder text2 = new StringBuilder("This document contains text and images. ");
        StringBuilder image1 = new StringBuilder("image1.png");
        doc.addtext(text1);
        doc.addtext(text2);
        doc.addimage(image1);
        doc.changetexttobold(0,1);
        doc.save();
    }
}